### DISCUSSION TITLE
Enable compression using middleware

### DISCUSSION THREAD

**Discussion by gkielwasser** [directus/directus#12973] [Archive]
Hi,

I'm trying to enable gzip compression.

I have done it this way:

```js
const compression = require('compression')
init('app.before', async function ({ app }) {
   if (process.env.GZIP_COMPRESSION === 'true') {
      app.use(compression())
   }
)
```

I notice it is not working on every requests.

- my custom endpoints : working
- directus endpoints (/items): working
- directus assets css/js: working
- directus endpoints (/refresh, /auth): **not working**
- directus logo image: working
- directus serving uploaded images: **not working**

What could be the reason some requests skip this compression ?

**Comment by rijkvanzanten** [comment]
Not sure! The `app.before` hook is called before any other middleware is registered, so it should work as expected. For refresh/auth, maybe it's because those are POST requests rather than GET?

**Reply by gkielwasser** [reply]
You right I have added a trace and I'm well going into the compression function for each requests.
```
app.use(compression({ filter: shouldCompress }))
```
Then its hope to the default filter function to decide wheter or not it should compress.

@azrikahar 
I have seen even with `content-type: application/json` POST request are not compressed. As you state, it might be the payload size that is too small.
Also I have seen `Content-Type: image/jpeg` is not compressed so this type might not be covered.

I didn't understood well all the criterias but it is should not anymore be related to Directus.

For now I will keep the default filter comptression strategy.

**Comment by azrikahar** [comment]
Just to add on two other possibilities I can think of, but this holds true more so for compressions via reverse proxies (at least from my personal experience) so they may not apply here:

- For the uploaded images, the default list of Content/file type to be compressed may not cover all, but typically this should be configurable

- Usually there is a minimum size on payload to compress, so it could be why refresh or auth payloads aren't compressed as they can be pretty small (compared to /items payload with multiple items)

**Comment by andrii-bodnar** [comment]
Hi @gkielwasser, 

I tried to set up this hook but I'm not sure if it's working...

`extensions/hooks/compression/index.js`:

```javascript
import compression from 'compression';

export default ({ init }) => {
    init('app.before', async function ({ app }) {
        app.use(compression());
    });
}
```

The response size is the same as before. Also, there is no `Content-Encoding` header in Directus responses.

Could someone suggest what I'm doing wrong? Thank you!

**Reply by gkielwasser** [reply]
Hi,

I do not see anything wrong, but as I stated in my previous post I didn't managed to have it working for all kind of responses (images for instance).

At least you should properly see the `content-encoding` response header in the GET requests.

**Reply by andrii-bodnar** [reply]
@gkielwasser thanks for the response!

Actually, I can't see this header in any response. Maybe, I should somehow additionally configure my Directus instance?

**Comment by andrii-bodnar** [comment]
Hi @gkielwasser, just updated my hook:

```javascript
const compression = require('compression');

init('app.before', async function ({ app }) {
    app.use(compression());
})
```

After starting the container with Directus, I can see the following logs:

```bash
...
- /directus/node_modules/directus/cli.js
    err: {
      "type": "Error",
      "message": "Cannot find module 'compression'\nRequire stack:\n- /directus/extensions/hooks/compression/index.js\n- /directus/node_modules/directus/dist/extensions.js\n- /directus/node_modules/directus/dist/services/graphql/index.js\n- /directus/node_modules/directus/dist/services/index.js\n- /directus/node_modules/directus/dist/middleware/respond.js\n- /directus/node_modules/directus/dist/controllers/activity.js\n- /directus/node_modules/directus/dist/app.js\n- /directus/node_modules/directus/dist/server.js\n- /directus/node_modules/directus/dist/cli/index.js\n- /directus/node_modules/directus/dist/cli/run.js\n- /directus/node_modules/directus/cli.js",
      "stack":
          Error: Cannot find module 'compression'
          Require stack:
          - /directus/extensions/hooks/compression/index.js
          - /directus/node_modules/directus/dist/extensions.js
          - /directus/node_modules/directus/dist/services/graphql/index.js
          - /directus/node_modules/directus/dist/services/index.js
          - /directus/node_modules/directus/dist/middleware/respond.js
          - /directus/node_modules/directus/dist/controllers/activity.js
          - /directus/node_modules/directus/dist/app.js
          - /directus/node_modules/directus/dist/server.js
          - /directus/node_modules/directus/dist/cli/index.js
          - /directus/node_modules/directus/dist/cli/run.js
          - /directus/node_modules/directus/cli.js
              at Function.Module._resolveFilename (node:internal/modules/cjs/loader:985:15)
              at Function.Module._load (node:internal/modules/cjs/loader:833:27)
              at Module.require (node:internal/modules/cjs/loader:1057:19)
              at require (node:internal/modules/cjs/helpers:103:18)
              at Object.<anonymous> (/directus/extensions/hooks/compression/index.js:1:21)
              at Module._compile (node:internal/modules/cjs/loader:1155:14)
              at Object.Module._extensions..js (node:internal/modules/cjs/loader:1209:10)
              at Module.load (node:internal/modules/cjs/loader:1033:32)
              at Function.Module._load (node:internal/modules/cjs/loader:868:12)
              at Module.require (node:internal/modules/cjs/loader:1057:19)
      "code": "MODULE_NOT_FOUND",
...
```

What should I do to enable this module for Directus?

**Reply by br41nslug** [reply]
You should be using the [Directus Extension SDK](https://docs.directus.io/extensions/creating-extensions.html) which will allow you to include this dependency in your extension. While it is possible to install the dependency in the root directus context using this CommonJS approach it is not a recommended way to go.

**Reply by andrii-bodnar** [reply]
@br41nslug thanks for the response!

I've found the extension for responses compression built by Directus Extensions SDK - [directus-extension-gzip-hook](https://github.com/liip/directus-extension-gzip-hook)

But I'm using Directus as a Docker container and I don't understand how I can install this extension in this case

**Reply by br41nslug** [reply]
Should be the same for docker as local npm installations, build and put it in the right location (a mounted volume for docker) https://docs.directus.io/extensions/creating-extensions.html#local-extension

**Reply by andrii-bodnar** [reply]
@br41nslug I should build the extension above and just use the resulting `index.js`?

**Reply by andrii-bodnar** [reply]
@br41nslug actually, it works for me, now responses are compressing, thanks!
